### ISSUE TITLE
using pry breaks requests

### ISSUE THREAD

**Issue by akostadinov** [loco2/lolsoap#17]
Hello, I am trying to use lolsoap for making webservice calls. I'm calling some stuff in a `pry` session. I constructed a login call to one service but it always failed. 
I then noticed in the body, there is always an extra element:
```
<ns0:pretty_print>#&lt;Pry::ColorPrinter:0x0000000a689228&gt;</ns0:pretty_print>
```

I nailed things down to pry printing the `LolSoap::Builder` object. If that is avoided, then things work. Please see for reproducer and more details this `pry` bug:
https://github.com/pry/pry/issues/1546

Also there is an ugly workaround that can be put into `losoap` source to avoid the issue.

On the one hand I think `pry` is doing too much by monkey patching just to print an object. This is my rational to file the `pry` bug. But also I'm wondering why are these `udef_method` calls in `lolsoap` and can they be avoided:
https://github.com/loco2/lolsoap/blob/a1805e3/lib/lolsoap/builder.rb#L22

Without these calls thing seems to be working fine. Also adding the `pry` specific methods as exceptions helps fix the issue.

Can things be handled more resiliently by `lolsoap`? Can we add `pry` specific workaround so that this useful tool works ok with `lolsoap`? I think `pry` has things to fix but they are non-trivial. And given popularity and usefulness of the tool, maybe a workaround in `lolsoap` is worth it?

**Comment by akostadinov** [comment]
reproducer, run this within `pry` and it adds extra elements, run without `pry` and it works:
```
    ## gem install lolsoap
    require "net/http"
    require "uri"
    require 'lolsoap'
     
    wsdl = URI.parse('http://almdemo.polarion.com/polarion/ws/services/SessionWebService?wsdl')
    session = LolSoap::Client.new(Net::HTTP.get_response(wsdl).body)
    login = session.request('logIn')
    login.body do |b|
      b.userName "USER"
      b.password "PASSWORD"
    end
    puts login.content
```

**Comment by akostadinov** [comment]
Fortunately it only happens when run command by command within `pry`. I guess it is related to `pry` pretty printing the result of each expression.

**Comment by banister** [comment]
@akostadinov you just say the call failed but you don't provide a backtrace, could you provide a backtrace for the failure?

**Comment by banister** [comment]
@akostadinov also, `pretty_print` is not from Pry, it comes from Ruby's own standard library. Pry isn't responsible for that monkey-patch, it's the standard library.  http://ruby-doc.org/stdlib-2.0.0/libdoc/pp/rdoc/PP.html

**Comment by akostadinov** [comment]
@banister , the call doesn't raise. It just inserts extra XML elements and server rejects it (returning an error which is not important). Please see linked /pry/pry/issues/1546 issue. I have pasted a lot of data there and didn't want to duplicate everything here for readability.

wrt method, are you able to replicate issue without pry?
Also there are 2 extra methods that are inserted - `instance_eval` and `instance_variables` (as described in the linked issue). Or is it the case that these methods are just restored after initially being deleted when pry tries to print object? In such case it makes even more sense to add said methods into the RESERVED list so they are never deleted and don't cause issues. WDYT?

Thank you for looking!

**Comment by akostadinov** [comment]
I have reproduced without involving `pry`. So it is obviously not a pry bug. Here it is:
```
require 'pp'
require "net/http"
require "uri"
require 'lolsoap'

wsdl = URI.parse('http://almdemo.polarion.com/polarion/ws/services/SessionWebService?wsdl')
session = LolSoap::Client.new(Net::HTTP.get_response(wsdl).body)
login = session.request('logIn')
pp(
  login.body do |b|
    b.userName "USER"
    b.password "PASSWORD"
  end
)
puts login.content
```

This would result in:
```

<?xml version="1.0" encoding="UTF-8"?>
<soap:Envelope xmlns:ns0="http://ws.polarion.com/SessionWebService-impl" xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Header/>
  <soap:Body>
    <ns0:logIn>
      <ns0:userName>USER</ns0:userName>
      <ns0:password>PASSWORD</ns0:password>
      <ns0:pretty_print>#&lt;PP:0x0000000117d3f0&gt;</ns0:pretty_print>
    </ns0:logIn>
  </soap:Body>
</soap:Envelope>
```

Where `ns0:pretty_print` is breaking the SOAP call.

**Comment by akostadinov** [comment]
Basically the issue is that `Builder` class removes most defined methods to avoid interference when creating request elements. But pretty print assumes some methods do exist and calls them on the `Builder`. That results in extra elements being added to request. I have files a PR #19 to avoid that and allow `pry` shell at least not break `Builder` just by creating it within the shell. If one tries more complicated things in `pry` shell with a `Builder` object, then that would probably also result in unexpected elements being added and other errors. But at least the most basic stuff should be fine now.

I don't think we can get any better except by stop undefining any pre-defined methods or the `Builder` class. This is not a bad idea actually but user would need to have a documented way to add elements when they interfere with existing method names.

IMO ideally lolsoap would have an API change and elements should be added like:
```
request.body do |b|
  b.element = "asdads"
end
```

It would make much more sense when reading the code as well will avoid the need to remove all kind of standard methods.

Alternatively the approach of `savon` can be taken and when elements match existing method, ask user to define like `b.element_ = "asdasd"`. But this is IMO uglier and requires more care from user.

Or stop removing standard methods but allow both styles `b.element "asdasd"` as well `b.element = "asdasd"`. This will provide basic backward compatibility (when users do not use overlapping method names) and allow saner user experience going forward.

My vote goes for the latter solution (stop undefining methods and allow both styles for adding elements).

Regards.

**Comment by eugenebolshakov** [comment]
Thanks, this is now fixed in 6d57e3f617fdc0519a4881a0089af6a34249b599

Regarding undefining methods, I think the author's idea behind that decision was to make `LolSoap::Builder` predictable in terms of what XML it generates and this issue with PrettyPrint kind of illustrates that point. PrettyPrint adds methods to all objects and other libraries can do too. Therefore it's possible that at some point you could start using a library that defines a method on all objects that has the same name as one of the elements in your SOAP request. Builder would then start to call that method instead of adding an XML element and so you'd end up with a different XML request. 

Hope my explanation makes sense.
